[llvm-lit] Fix Unhashable TypeError when using lit's internal shell

Harini0924 opened the pull request:
When using the lit internal shell with the command:
```
LIT_USE_INTERNAL_SHELL=1 ninja check-compiler-rt
```
The follow error is encountered:
```
File "TestRunner.py", line 770, in _executeShCmd
    inproc_builtin = inproc_builtins.get(args[0], None)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: unhashable type: 'GlobItem'
```
This error is in a compiler-rt file:
```
TestCases/Linux/long-object-path.cpp
```
This error occurs because `args[0]` is of type `GlobItem`, which is not hashable, leading to a `TypeError` when it is passed in `inproc_builtins.get()`. To resolve this issue,  I have updated the implementation to ensure that `args[0]` is hashable before it is used in `inproc_builtins`.
fixes: #102389
[link to RFC](https://discourse.llvm.org/t/rfc-enabling-the-lit-internal-shell-by-default/80179)

Harini0924 committed:
[llvm-lit] Unhashable type Support

Used expand_glob_expressions to resolve all GlobItem instances in args
list to their acutal paths. Checked and converted args[0] if it is a
Glob instance and ensured the resolved path is a string (hashable) before
using it as a key in the inproc_builtins dictionary. To sum
everything, this patch eliminates the 'unhashable type: GlobItem' error.

Harini0924 commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -767,6 +767,13 @@ def _executeShCmd(cmd, shenv, results, timeoutHelper):
+
+        # Ensure args[0] is hashable before using it in inproc_builtins
+        if isinstance(args[0], GlobItem):
+            expanded_args = expand_glob(args[0], cmd_shenv.cwd)
```
I wanted to get your feedback on the recent changes I made. Specifically, I added a check to ensure args[0] is hashable before using it in inproc_builtins. If args[0] is an instance of GlobItem, I'm expanding it to its full path. I wasn't sure if this approach is the most efficient or if there's a better way to resolve the GlobItem TypeError. Any insights or suggestions for improvement would be greatly appreciated!

ilovepi replied:
This isn't a check if it's `Hashable` though, it's just a check if the instance is a `GlobItem`.

Why not just do the expansion at the call site though? `expand_glob()` is already checking if it's a `GlobItem`. It also won't modify `args[0]`, which seems preferable, since `expand_glob_expressions()` goes out of its way not to modify `args[0]`.
```
inproc_builtin = inproc_builtins.get(expand_glob(args[0], cmd_shenv.cwd)[0], None)
```

Harini0924 replied:
I have updated the implementation to ensure `args[0]` is hashable before using it in inproc_builtins. The previous code:
```
# Ensure args[0] is hashable before using it in inproc_builtins
if isinstance(args[0], GlobItem):
    expanded_args = expand_glob(args[0], cmd_shenv.cwd)
    if expanded_args:
        args[0] = expanded_args[0]
```
has been modifed to: 
```
# Ensure args[0] is hashable
args[0] = expand_glob(args[0], cmd_shenv.cwd)[0]
```
You raised a valid point about performing the expansion at the call site, as `expand_glob()` already checks if `args[0]` is a `GlobItem` and does not modify `args[0]`, which aligns with the approach taken by `expand_glob_expressions()`. However, the current error `TypeError: unhashable type: 'GlobItem'` occurs on the line:
```
inproc_builtin = inproc_builtins.get(args[0], None)
```
After changing this line to:
```
inproc_builtin = inproc_builtins.get(expand_glob(args[0], cmd_shenv.cwd)[0], None)
```
the same `TypeError` occurs on subsequent lines where `args[0]` is called. Instead of repeatedly applying the same functionality to `args[0]` until line 866 where args is expanded with:
```# Expand all glob expressions
args = expand_glob_expressions(args, cmd_shenv.cwd)
```
it would be more efficient to expand args[0] before invoking any in-process built-ins. This approach resolves the `TypeError` for this and all subsequent lines where `args[0]` is used.

By making this change, we address the issue holistically, ensuring the robustness of the implementation against unhashable types like `GlobItem`.

Let me know if this approach works for you or if you have any further suggestions.

Harini0924 commented:
CC: @ilovepi @petrhosek

ilovepi reviewed (commented):
Please update the PR title to reflect what you are fixing/doing. This PR doesn't add support for Unhashable TypeErrors, it's trying to fix the manifestation of a TypeError in lit's python code. Additionally, please update the description to describe what you are fixing, and summarize the core changes you're making.

In this case, when using the internal shell, you're fixing a type mismatch when using certain builtins, because the GlobItem type doesn't implement hashable. As a workaround you've expanded the glob to avoid the error without additional code changes.

ilovepi commented:
Suggested title:
`[llvm-lit] Fix Unhashable TypeError when using lit's internal shell`

For the description:
```
When using lit's internal shell, commands like ______ result in a type error. This happens because when
____ is called in TestRunner.py its passed a GlobItem, which does not implement Hashable.

We could either implement the required interface, or more simply, ensure that the passed in parameter is of the correct type. expand_glob is used elsewhere to convert GlobItems into _____ types, which are hashable, and is used at the call site to prevent this error. ...
```

You can fill in more details, but this is a more useful type of commit message, and communicates a lot to other developers coming along later. Also, you don't need to use my text verbatim (er even something so long winded!), but something more along these lines would be a big improvement in my opinion.

If you look through the commit logs you'll see there isn't a strong convention in how people describe their patches. Some will be very short, some will be very long, and most somewhere between. However, I'd also encourage you to scan the patch, and see what it does for yourself. After a brief survey, I think you'll agree that when authors provide a sufficiently detailed account of what they did and why, its usually easier to follow the important bits in the patch.

Harini0924 committed:
[llvm-lit] Fix handling of GlobItem in args[0] to ensure it is hashable

Removed the previous complex implementation to a simipler verison.
Added direct expansion and conversion of args[0] to ensure it is
always a resolved path string. This patch ensures that args[0] is
consistently a string, preventing TypeError when args[0] is used as
a dictionary key or in other contexts that require a hashable type.

Harini0924 committed:
Merge branch 'main' into fixing_unhasable

Harini0924 committed:
[llvm-lit] Simplify args[0] Hashability Check and Expansion

Refactored the code to simplify the hashability check and expansion
of args[0]. The pervious commit included a conditional check to
determine if args[0] was a GlobItem and subsequently expanded it.
The new implementation directly assigns the expanded value of args[0].
This patch simplifies the logic by removing unnecessary conditionals
and ensures that args[0] is always assigned an expanded value.

ilovepi reviewed (approved):
Given the fact that only selectively expanding `arg[0]` seems to be causing more expansions and fix ups, I'm OK with the current approach. 

However,  the description is still unclear, despite the template I provided. 

Please include:
1. an example of the command that triggers this error
2. at least one example of where this happens in our codebase
1. conditions for running the test w/ the internal shell 
2. remove "This patch eliminates the 'unhashable type: GlobItem' error." (at most this stops 1 specific instance from occurring).

These details will help others coming by later understand why you're making this change. It's also worth including a link to the RFC to let others know this is part of that effort/cleanup.

jh7370 commented on llvm/utils/lit/lit/TestRunner.py:
```diff
@@ -767,6 +767,10 @@ def _executeShCmd(cmd, shenv, results, timeoutHelper):
         # FIXME: Standardize on the builtin echo implementation. We can use a
         # temporary file to sidestep blocking pipe write issues.
+
+        # Ensure args[0] is hashable
```
```suggestion
        # Ensure args[0] is hashable.
```
Nit

Harini0924 committed:
[llvm-lit] Fixed a comment

Added a period to the end of my comment.

ilovepi reviewed (approved):
LGTM.
